File: backend/coverage/lawnmower.py
import math
from typing import List, Tuple, Dict

LatLng = Dict[str, float]

def _to_local(vertices: List[LatLng]) -> List[Tuple[float,float]]:
    """Project LatLng list to local metre plane anchored at vertices[0]."""
    o = vertices[0]
    mplat = 110540.0
    mplng = 111320.0 * math.cos(math.radians(o['lat']))
    return [(( v['lng'] - o['lng']) * mplng,
              (v['lat'] - o['lat']) * mplat) for v in vertices]

def _from_local(x: float, y: float, origin: LatLng) -> LatLng:
    mplat = 110540.0
    mplng = 111320.0 * math.cos(math.radians(origin['lat']))
    return {'lat': origin['lat'] + y / mplat, 'lng': origin['lng'] + x / mplng}

def _scanline_intersections(poly: List[Tuple[float,float]], scan_y: float) -> List[float]:
    xs = []
    n = len(poly)
    for i in range(n):
        ax, ay = poly[i]
        bx, by = poly[(i+1) % n]
        lo, hi = min(ay, by), max(ay, by)
        if scan_y < lo or scan_y >= hi:
            continue
        t = (scan_y - ay) / (by - ay)
        xs.append(ax + t * (bx - ax))
    return sorted(xs)

def generate_lawnmower(polygon: List[LatLng], sweep_width_m: float) -> List[LatLng]:
    """
    Boustrophedon (lawnmower) coverage path for an arbitrary polygon.
    Returns ordered LatLng waypoints covering the full interior.
    """
    if len(polygon) < 3:
        return []
    sw = max(sweep_width_m, 1.0)
    origin = polygon[0]
    poly = _to_local(polygon)

    ys = [p[1] for p in poly]
    y_min, y_max = min(ys), max(ys)

    result: List[LatLng] = []
    row = 0
    y = y_min + sw / 2
    while y <= y_max:
        xs = _scanline_intersections(poly, y)
        for p in range(0, len(xs) - 1, 2):
            xl, xr = xs[p], xs[p+1]
            left  = _from_local(xl, y, origin)
            right = _from_local(xr, y, origin)
            if row % 2 == 0:
                result.extend([left, right])
            else:
                result.extend([right, left])
            row += 1
        y += sw
    return result

File: simulator/src/engine/lawnmower.ts
// Ported from backend/coverage/lawnmower.py — generate_lawnmower
// Exact math preserved: same projection, same scanline intersection logic.

import { METERS_PER_DEG_LAT, metersPerDegLng } from './geo';
import type { LatLng } from '../types';

function toLocal(vertices: LatLng[]): [number, number][] {
  // Port of lawnmower.py::_to_local
  const o = vertices[0];
  const mPerLng = metersPerDegLng(o.lat);
  return vertices.map(v => [
    (v.lng - o.lng) * mPerLng,
    (v.lat - o.lat) * METERS_PER_DEG_LAT,
  ]);
}

function fromLocal(x: number, y: number, origin: LatLng): LatLng {
  // Port of lawnmower.py::_from_local
  const mPerLng = metersPerDegLng(origin.lat);
  return {
    lat: origin.lat + y / METERS_PER_DEG_LAT,
    lng: origin.lng + x / mPerLng,
  };
}

function scanlineIntersections(poly: [number, number][], scanY: number): number[] {
  // Port of lawnmower.py::_scanline_intersections
  const xs: number[] = [];
  const n = poly.length;
  for (let i = 0; i < n; i++) {
    const [ax, ay] = poly[i];
    const [bx, by] = poly[(i + 1) % n];
    const lo = Math.min(ay, by);
    const hi = Math.max(ay, by);
    if (scanY < lo || scanY >= hi) continue;
    const t = (scanY - ay) / (by - ay);
    xs.push(ax + t * (bx - ax));
  }
  return xs.sort((a, b) => a - b);
}

/**
 * Port of lawnmower.py::generate_lawnmower
 * Boustrophedon coverage path for an arbitrary polygon.
 */
export function generateLawnmower(polygon: LatLng[], sweepWidthM: number): LatLng[] {
  if (polygon.length < 3) return [];
  const sw = Math.max(sweepWidthM, 1.0);
  const origin = polygon[0];
  const poly = toLocal(polygon);

  const ys = poly.map(p => p[1]);
  const yMin = Math.min(...ys);
  const yMax = Math.max(...ys);

  const result: LatLng[] = [];
  let row = 0;
  let y = yMin + sw / 2;

  while (y <= yMax) {
    const xs = scanlineIntersections(poly, y);
    for (let p = 0; p < xs.length - 1; p += 2) {
      const xl = xs[p];
      const xr = xs[p + 1];
      const left = fromLocal(xl, y, origin);
      const right = fromLocal(xr, y, origin);
      if (row % 2 === 0) {
        result.push(left, right);
      } else {
        result.push(right, left);
      }
      row++;
    }
    y += sw;
  }
  return result;
}
